worlds/messenger/__init__.py

```diff
@@ -54,58 +67,139 @@ class MessengerWorld(World):
+            "Corrupted Future - Key of Courage",
+        },
+        "Phobe": {
+            "Catacombs - Necro",
+            "Bamboo Creek - Claustro",
+            "Searing Crags - Pyro",
+            "Cloud Ruins - Acro",
+        },
+        "Phobekin": {
+            "Catacombs - Necro",
+            "Bamboo Creek - Claustro",
+            "Searing Crags - Pyro",
+            "Cloud Ruins - Acro",
+        },
+    }
 
-    required_client_version = (0, 4, 2)
+    required_client_version = (0, 4, 3)
 
     web = MessengerWeb()
 
     total_seals: int = 0
     required_seals: int = 0
+    created_seals: int = 0
     total_shards: int = 0
     shop_prices: Dict[str, int]
     figurine_prices: Dict[str, int]
     _filler_items: List[str]
+    starting_portals: List[str]
+    plando_portals: List[str]
+    spoiler_portal_mapping: Dict[str, str]
+    portal_mapping: List[int]
+    transitions: List[Entrance]
+    reachable_locs: int = 0
 
     def generate_early(self) -> None:
         if self.options.goal == Goal.option_power_seal_hunt:
-            self.options.shuffle_seals.value = PowerSeals.option_true
             self.total_seals = self.options.total_seals.value
 
+        if self.options.limited_movement:
+            self.options.accessibility.value = Accessibility.option_minimal
+            if self.options.logic_level < Logic.option_hard:
+                self.options.logic_level.value = Logic.option_hard
+
+        self.multiworld.early_items[self.player]["Meditation"] = self.options.early_meditation.value
+
         self.shop_prices, self.figurine_prices = shuffle_shop_prices(self)
 
+        starting_portals = ["Autumn Hills", "Howling Grotto", "Glacial Peak", "Riviere Turquoise", "Sunken Shrine", "Searing Crags"]
+        self.starting_portals = [f"{portal} Portal"
+                                 for portal in starting_portals[:3] +
+                                 self.random.sample(starting_portals[3:], k=self.options.available_portals - 3)]
+        self.plando_portals = []
+        self.portal_mapping = []
+        self.spoiler_portal_mapping = {}
+        self.transitions = []
+
+    @classmethod
+    def stage_generate_early(cls, multiworld: MultiWorld):
+        if multiworld.players > 1:
+            return
+        cls.generate_output = generate_output
+
     def create_regions(self) -> None:
         # MessengerRegion adds itself to the multiworld
-        for region in [MessengerRegion(reg_name, self) for reg_name in REGIONS]:
-            if region.name in REGION_CONNECTIONS:
-                region.add_exits(REGION_CONNECTIONS[region.name])
+        # create simple regions
+        for level in LEVELS:
+            MessengerRegion(level, self)
+        # create and connect complex regions that have sub-regions
+        for region in [MessengerRegion(f"{parent} - {reg_name}", self, parent)
+                       for parent, sub_region in CONNECTIONS.items()
+                       for reg_name in sub_region]:
```

Comment:
i don't like creating variables for things i'm only going to use a single time if i can reasonably go without. this is a case where i agree the comprehension shouldn't have been here, though i will be keeping it (because like you said i find it perfectly legible).